File: lib/common/eventhandlers.py
import logging

import adsk.core

from .uicontext import uicontext

handlers = []

logger = logging.getLogger('silvanus').getChild(__name__)


class CommandDestroyHandler(adsk.core.CommandEventHandler):

    def __init__(self, command):
        super().__init__()
        self.app = command.app
        self.command = command

    def notify(self, args):
        with uicontext(self.app):
            self.command.on_destroy(args)


class CommandExecuteHandler(adsk.core.CommandEventHandler):

    def __init__(self, command):
        super().__init__()
        self.command = command
        self.app = command.app

    def notify(self, args):
        with uicontext(self.app):
            command = args.firingEvent.sender
            inputs = command.commandInputs

            self.command.on_execute(inputs)


class InputChangedHandler(adsk.core.InputChangedEventHandler):

    def __init__(self, command):
        super().__init__()
        self.command = command
        self.app = command.app

    def notify(self, args):
        with uicontext(self.app):
            cmd_input = args.input

            self.command.on_change(cmd_input)


class ExecutePreviewHandler(adsk.core.CommandEventHandler):

    def __init__(self, command, preview_id):
        super().__init__()
        self.command = command
        self.app = command.app
        self.preview_id = preview_id

    def notify(self, args):
        with uicontext(self.app):
            command = args.firingEvent.sender
            inputs = command.commandInputs

            self.command.on_preview(inputs)


class ValidateInputsHandler(adsk.core.ValidateInputsEventHandler):

    def __init__(self, command):
        super().__init__()
        self.command = command
        self.app = command.app

    def notify(self, args):
        with uicontext(self.app):
            command = args.firingEvent.sender
            inputs = command.commandInputs

            args.areInputsValid = self.command.on_validate(inputs)


class CreatedEventHandler(adsk.core.CommandCreatedEventHandler):

    def __init__(self, command, preview_id):
        super().__init__()
        self.command = command
        self.app = self.command.app
        self.preview_id = preview_id

    def notify(self, args):
        global handlers

        with uicontext(self.app):
            command = args.command

            on_execute = CommandExecuteHandler(self.command)
            command.execute.add(on_execute)
            handlers.append(on_execute)

            on_preview = ExecutePreviewHandler(self.command, self.preview_id)
            command.executePreview.add(on_preview)
            handlers.append(on_preview)

            on_change = InputChangedHandler(self.command)
            command.inputChanged.add(on_change)
            handlers.append(on_change)

            on_validate = ValidateInputsHandler(self.command)
            command.validateInputs.add(on_validate)
            handlers.append(on_validate)

            on_destroy = CommandDestroyHandler(self.command)
            command.destroy.add(on_destroy)
            handlers.append(on_destroy)

            self.command.on_create(command)

File: src/lib/common/eventhandlers.cpp
//
//  SilvanusPro
//
//  Created by Hobbyist Maker on 7/21/20.
//  Copyright © 2020 HobbyistMaker. All rights reserved.
//

#include "eventhandlers.hpp"

#include <Core/CoreAll.h>

using namespace adsk::core;
using namespace adsk::fusion;

using namespace silvanus::common;

void CreatedEventHandler::notify(const Ptr<CommandCreatedEventArgs>& event_args) {
    auto command = event_args->command();
    if (!command)
        return;
    
    auto on_destroy = Ptr<CommandEvent>{command->destroy()};
    auto is_ok = on_destroy && on_destroy->add(&on_destroy_handler) | true;

    auto on_execute = Ptr<CommandEvent>{command->execute()};
    is_ok = (is_ok & on_execute) != 0 && on_execute->add(&on_execute_handler) | true;

    auto on_change = Ptr<InputChangedEvent>{command->inputChanged()};
    is_ok = (is_ok & on_change) != 0 && on_change->add(&on_change_handler) | true;

    auto on_validate = Ptr<ValidateInputsEvent>{command->validateInputs()};
    is_ok = (is_ok & on_validate) != 0 && on_validate->add(&on_validate_handler) | true;

    auto on_preview = Ptr<CommandEvent>{command->executePreview()};
    is_ok = (is_ok & on_preview) != 0 && on_preview->add(&on_preview_handler) | true;

    if (!is_ok)
        return;

    m_command.lock()->create(event_args);
}
